jupyter_sparkmonitor/sparkmonitor.sh

```diff
@@ -0,0 +1,132 @@
+readonly EXISTING_PYSPARK_KERNEL='pyspark'
+
+function retry_apt_command() {
+  cmd="$1"
+  for ((i = 0; i < 10; i++)); do
+    if eval "$cmd"; then
+      return 0
+    fi
+    sleep 5
+  done
+  return 1
+}
+
+function update_apt_get() {
+  retry_apt_command "apt-get update"
+}
+
+function install_apt_get() {
+  pkgs="$@"
+  retry_apt_command "apt-get install -y $pkgs"
+}
+
+function err() {
+  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $@" >&2
+  return 1
+}
+
+function install_sparkmonitor(){
+  # Install asyncio
+  retry_apt_command "${CONDA_DIRECTORY}/bin/pip install asyncio" ||
+    err "Failed to install asyncio"
+
+  # Remove existing PySpark kernel as it uses pyspark shell instead of ipython
+  retry_apt_command "${CONDA_DIRECTORY}/bin/jupyter kernelspec remove ${EXISTING_PYSPARK_KERNEL} -f" ||
+    err "Failed to remove existing PySpark kernel"
+
+  # Install sparkmonitor. Don't mind if it fails to start the first time.
+  retry_apt_command "/${CONDA_DIRECTORY}/bin/pip install sparkmonitor"
+  if [ $? != 0 ]; then
+    err 'Failed to install sparkmonitor'
+  fi
+}
+
+function configure_sparkmonitor(){
+  local sparkmonitor_version;
+  sparkmonitor_version="$(${CONDA_DIRECTORY}/bin/pip list | grep -i sparkmonitor | awk 'END {print $2}')"
+  retry_apt_command "${CONDA_DIRECTORY}/bin/jupyter nbextension install sparkmonitor --py --system --symlink"
+  retry_apt_command "${CONDA_DIRECTORY}/bin/jupyter nbextension enable sparkmonitor --py --system"
+  retry_apt_command "${CONDA_DIRECTORY}/bin/jupyter serverextension enable --py --system sparkmonitor"
+  retry_apt_command "${CONDA_DIRECTORY}/bin/ipython profile create"
+  local ipython_profile_location="$(${CONDA_DIRECTORY}/bin/ipython profile locate default)"
+  echo "c.InteractiveShellApp.extensions.append('sparkmonitor.kernelextension')" >>  ${ipython_profile_location}/ipython_kernel_config.py
+}
+
+function install_pyspark_kernel() {
+  echo "Installing pyspark Kernel..."
+  local pyspark_kernel_tmp_dir=$(mktemp -d -t dataproc-init-actions-sparkmonitor-XXXX)
+
+  cat << EOF >"${pyspark_kernel_tmp_dir}/kernel.json"
+  {
+   "argv": [
+      "python", "-m", "ipykernel", "-f", "{connection_file}"],
+   "display_name": "PySpark",
+   "language": "python",
+   "env": {
+      "SPARK_HOME": "/usr/lib/spark/",
+      "PYTHONPATH": "${PYTHON_PATH}"
+   }
+  }
+EOF
+
+  /opt/conda/default/bin/jupyter kernelspec install "${pyspark_kernel_tmp_dir}"
+  rm -rf $pyspark_kernel_tmp_dir
+  echo "PySpark kernel setup completed!"
+}
+function main() {
+  if [[ ! -f "${JUPYTER_INIT_SCRIPT}" ]]; then
+    err "Jupyter component is missing"
+    exit 1
+  fi
+
+  if [[ ! -d "${CONDA_DIRECTORY}" ]]; then
+    err "Anaconda component is missing"
+    exit 1
+  fi
+
+  if [[ -f /etc/profile.d/conda.sh ]]; then
+    source /etc/profile.d/conda.sh
+  fi
+
+  if [[ -f /etc/profile.d/effective-python.sh ]]; then
+    source /etc/profile.d/effective-python.sh
+  fi
+
+  if [[ "${ROLE}" == 'Master' ]]; then
```

Comment:
This init action only needs to run on the master, right? I think you can just `exit 0` from the top of `main` if it's not the master